include/nbl/asset/filters/kernels/WeightFunctions.h

```diff
@@ -0,0 +1,308 @@
+			}
+			else if constexpr (derivative == 1)
+			{
+				const auto PIx = core::PI<double>() * x;
+				double f = core::sinc(PIx);
+				double df = core::PI<double>() * core::d_sinc(PIx);
+				double g = core::KaiserWindow(x, alpha, absMinSupport);
+				double dg = core::d_KaiserWindow(x, alpha, absMinSupport);
+				return df * g + f * dg;
+			}
+			else
+			{
+				static_assert(false, "TODO");
+				return core::nan<double>();
+			}
+		}
+		return 0.0;
+	}
+};
+
+// This is the interface for 1D weight functions that can be used to create a `CFloatingPointSeparableImageFilterKernel`.
+// Current implementations of this interface are:
+// - `CWeightFunction1D`
+// - `CChannelIndependentWeightFunction1D`
+// - `CConvolutionWeightFunction1D`
+template<typename T>
+concept KernelWeightFunction1D = requires(T t, const float x, const uint8_t channel)
+{
+	{ t.weight(x, channel) }	-> std::same_as<double>;
+	{ t.getMinSupport() }		-> std::same_as<float>;
+	{ t.getMaxSupport() }		-> std::same_as<float>;
+	{ t.stretch(x) }			-> std::same_as<void>;
+	{ t.scale(x) }				-> std::same_as<void>;
+	{ t.stretchAndScale(x) }	-> std::same_as<void>;
+};
+
+template <typename Function1D, int32_t derivative = 0>
+class CWeightFunction1D /*final*/ // TODO(achal): Cannot make it final just yet because `weight_function_value_type` inherits from this.
+{
+	static_assert(derivative <= Function1D::k_smoothness);
+public:
+	constexpr static inline uint32_t k_smoothness = Function1D::k_smoothness-derivative;
+	constexpr static inline float k_energy[4] = { 0.f, 0.f, 0.f, 0.f }; // TODO(achal): Implement.
+
+	// Calling: f(x).stretch(2) will obviously give you f(x/2)
+	inline void stretch(const float s)
+	{
+		assert(s != 0.f);
+
+		m_minSupport *= s;
+		m_maxSupport *= s;
+
+		const float rcp_s = 1.f / s;
+
+		m_invStretch *= rcp_s;
+
+		if constexpr (derivative != 0)
+			scale(pow(rcp_s, derivative));
+	}
+
+	inline void scale(const float s)
+	{
+		assert(s != 0.f);
+		m_totalScale *= s;
+	}
+
+	// This method will keep the integral of the weight function constant.
+	inline void stretchAndScale(const float stretchFactor)
+	{
+		stretch(stretchFactor);
+		scale(1.f / stretchFactor);
+	}
+
+	inline double weight(const float x, const uint8_t channel) const
+	{
+		return static_cast<double>(m_totalScale*Function1D::weight<derivative>(x*m_invStretch));
+	}
+
+	inline float getMinSupport() const { return m_minSupport; }
+	inline float getMaxSupport() const { return m_maxSupport; }
+	inline float getInvStretch(const uint32_t channel = 0) const { return m_invStretch; }
+	inline float getTotalScale(const uint32_t channel = 0) const { return m_totalScale; }
+
+private:
+	float m_minSupport = Function1D::min_support;
+	float m_maxSupport = Function1D::max_support;
+	float m_invStretch = 1.f;
+	float m_totalScale = 1.f;
+};
+
+namespace impl
+{
+
+template<class WeightFunction1D>
+struct weight_function_value_type : protected WeightFunction1D
```

Comment:
you don't need to inherit to access `weight` via declval

unless some weight function's `weight()` is protected?

Reply:
which it isn't

Reply:
I'd get rid of this whole namespace and declare a 
```cpp
public:
using value_type = decltype(std::declval<Function1D>().weight(0.f));
```
inside of the `CWeightFunction1D`

and then use that to express stretch, scale, `k_energy` and the return value type

Reply:
done by myself.